hail/src/main/scala/is/hail/table/Table.scala

```diff
@@ -27,22 +28,65 @@ case object Descending extends SortOrder
 
 case class SortField(field: String, sortOrder: SortOrder)
 
+case class IndexSpec(rel_path: String, offset_field: Option[String] = None)
+
 abstract class AbstractTableSpec extends RelationalSpec {
   def references_rel_path: String
   def table_type: TableType
+  def index_spec: Option[IndexSpec]
   def rowsComponent: RVDComponentSpec = getComponent[RVDComponentSpec]("rows")
+
+  def rowsSpec(path: String): AbstractRVDSpec = AbstractRVDSpec.read(HailContext.get, path + "/" + rowsComponent.rel_path)
   def rvdType(path: String): RVDType = {
-    val rows = AbstractRVDSpec.read(HailContext.get, path + "/" + rowsComponent.rel_path)
+    val rows = rowsSpec(path)
     RVDType(rows.encodedType, rows.key)
   }
+
+  def readIndexed(
+    hc: HailContext,
+    path: String,
+    rowType: PStruct,
+    partitioner: Option[RVDPartitioner],
+    filterIntervals: Boolean): RVD
 }
 
 case class TableSpec(
   file_version: Int,
   hail_version: String,
+  index_spec: Option[IndexSpec],
   references_rel_path: String,
   table_type: TableType,
-  components: Map[String, ComponentSpec]) extends AbstractTableSpec
+  components: Map[String, ComponentSpec]) extends AbstractTableSpec {
+
+  def readIndexed(
+    hc: HailContext,
+    path: String,
+    rowType: PStruct,
+    partitioner: Option[RVDPartitioner],
+    filterIntervals: Boolean
+  ): RVD = {
+    if (index_spec.isEmpty)
+      fatal(s"attempted to read with index from an unindexed table: $path")
+
+    val requestedKey = table_type.key.takeWhile(rowType.hasField)
+    val rows = rowsSpec(path).asInstanceOf[OrderedRVDSpec]
```

Comment:
I don't think you can do this cast

Reply:
I can definitely do this cast. All `1.1.0` table's `rowsSpec`s are `OrderedRVDSpec`s. I need access to the partitioner.

Reply:
Won't casting to an AbstractRVDSpec also work? I do realize this will work right now, but that seems safer (if for instance, we upgrade the file format without changing the index stuff)